pootle/apps/virtualfolder/utils.py

```diff
@@ -0,0 +1,220 @@
+
+    @property
+    def matching_stores(self):
+        """Store qs containing all stores that match
+        project, language, and rules for this vfolder
+        """
+        return self.filter_from_rules(self.store_qs)
+
+    @property
+    def rules(self):
+        """Glob matching rules"""
+        return (
+            "%s" % r.strip()
+            for r
+            in self.vf.filter_rules.split(","))
+
+    @property
+    def store_manager(self):
+        """The root object manager for finding/adding stores"""
+        return Store.objects
+
+    @property
+    def store_qs(self):
+        """The stores qs without any rule filtering"""
+        return self.filter_projects(
+            self.filter_languages(
+                self.store_manager))
+
+    def add_and_remove_stores(self):
+        """Add Stores that should be associated but arent, delete Store
+        associations for Stores that are associated but shouldnt be
+        """
+        existing_stores = set(self.existing_stores)
+        matching_stores = set(self.matching_stores)
+        to_add = matching_stores - existing_stores
+        to_remove = existing_stores - matching_stores
+        if to_add:
+            self.add_stores(to_add)
+        if to_remove:
+            self.remove_stores(to_remove)
+        return to_add, to_remove
+
+    def add_stores(self, stores):
+        """Associate a Store"""
+        self.vf.stores.add(*stores)
+
+    def filter_from_rules(self, qs):
+        filtered_qs = qs.none()
+        for rule in self.rules:
+            filtered_qs = (
+                filtered_qs
+                | qs.filter(pootle_path__regex=self.get_rule_regex(rule)))
+        return filtered_qs
+
+    def filter_languages(self, qs):
+        if self.languages is None:
+            return qs
+        return qs.filter(
+            translation_project__language_id__in=self.languages)
+
+    def filter_projects(self, qs):
+        if self.projects is None:
+            return qs
+        return qs.filter(
+            translation_project__project_id__in=self.projects)
+
+    def get_rule_regex(self, rule):
+        """For a given *glob* rule, return a pootle_path *regex*"""
+        return (
+            "^%s%s"
+            % (self.tp_path,
+               PathFilter().path_regex(rule)))
+
+    def path_matches(self, path):
+        """Returns bool of whether path is valid for this VF.
+        """
+        for rule in self.rules:
+            if fnmatch(rule, path):
+                return True
+        return False
+
+    def remove_stores(self, stores):
+        self.vf.stores.remove(*stores)
+
+    def should_add_store(self, store):
+        if self.store_matches(store) and not self.store_associated(store):
+            return True
+        return False
+
+    def store_associated(self, store):
+        return self.vf.stores.through.objects.filter(
+            store_id=store.id,
+            virtualfolder_id=self.vf.id).exists()
```

Comment:
If `self.vf.stores.through` is just the same as `Store.objects` I would use the latter instead.